server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceStateMachineTests.java

```diff
@@ -0,0 +1,1132 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.SETTING_CREATION_DATE;
+import static org.elasticsearch.cluster.metadata.IndexMetadata.SETTING_INDEX_UUID;
+import static org.elasticsearch.snapshots.SnapshotsService.UPDATE_SNAPSHOT_STATUS_ACTION_NAME;
+import static org.hamcrest.Matchers.hasItem;
+
+public class SnapshotsServiceStateMachineTests extends ESTestCase {
+
+    private static final String REPOSITORY_TYPE = "fake";
+    private DeterministicTaskQueue deterministicTaskQueue;
+    private ThreadPool threadPool;
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        deterministicTaskQueue = new DeterministicTaskQueue();
+        deterministicTaskQueue.setExecutionDelayVariabilityMillis(1000);
+        threadPool = deterministicTaskQueue.getThreadPool();
+    }
+
+    public void testBasicActivities() {
+        final var repositoryName = "repo";
+        try (var ts = createServices(Settings.EMPTY)) {
+            final var future = new PlainActionFuture<Void>();
+            SubscribableListener
+
+                .<AcknowledgedResponse>newForked(
+                    l -> ts.repositoriesService.registerRepository(new PutRepositoryRequest(repositoryName).type(REPOSITORY_TYPE), l)
+                )
+
+                .<SnapshotInfo>andThen((l, r) -> {
+                    logger.info("--> create snapshot");
+                    ts.snapshotsService.executeSnapshot(new CreateSnapshotRequest(repositoryName, "snap-1"), l);
+                })
+
+                .<SnapshotInfo>andThen((l, r) -> {
+                    logger.info("--> create snapshot");
+                    ts.snapshotsService.executeSnapshot(new CreateSnapshotRequest(repositoryName, "snap-2"), l);
+                })
+
+                .<Void>andThen((l, r) -> {
+                    logger.info("--> clone snapshot");
+                    ts.snapshotsService.cloneSnapshot(
+                        new CloneSnapshotRequest(repositoryName, "snap-2", "snap-3", new String[] { "*" }),
+                        l
+                    );
+                })
+
+                .<Void>andThen((l, r) -> {
+                    final var snapshotNames = Set.of("snap-1", "snap-2", "snap-3");
+                    ts.snapshotsService.deleteSnapshots(
+                        new DeleteSnapshotRequest(repositoryName, randomArray(1, 2, String[]::new, () -> randomFrom(snapshotNames))),
+                        l
+                    );
+                })
+
+                // TODO: repository cleanup
```

Comment:
The code for repository cleanup is mostly located inside the transport action instead of SnapshotService. I think this might be something that we can improve?